### DISCUSSION TITLE
Howto: Using purgecss with vite and vanilla-extract

### DISCUSSION THREAD

**Discussion by Hecatron** [vanilla-extract-css/vanilla-extract#1132] [General]
Hi,
I've recently discovered an easy way to use purgecss with vanilla extract while using the vite bundler.
I figured this might be useful for others who might be using sprinkles.

## Short Version

The short version is to use this plugin within vite instead of using purgecss via postcss
and to make sure any call to a sprinkles function is done within a .css.ts file

  * https://github.com/mojojoejo/vite-plugin-purgecss

## Longer Version

One of the really handy features of vanilla-extract is to the use of sprinkles to create atomic css rules.
However this can result in a lot of the properties that are not used ending up in the final css bundle.
Resulting in very big css bundles

There's been some discussions about if tree shaking should be supported etc but currently that's not going to be a thing.
Rainbow Sprinkles uses a different approach via css variables

  * https://github.com/vanilla-extract-css/vanilla-extract/discussions/32
  * https://github.com/vanilla-extract-css/vanilla-extract/issues/700
  * https://github.com/wayfair/rainbow-sprinkles

One of the ways to remove redundent css is via purgecss, which I believe is a common approach when using tailwind.
However if you try to insert purgecss as a postcss plugin then provide it a directory via it's content property
```js
const purgecss = require('@fullhuman/postcss-purgecss')
module.exports = {
  plugins: [
    purgecss({
      // Doesn't work purges everything
      content: ['./src/**/*.vue']
    })
  ]
}
```
This isn't going to work, with the above more than likley purgecss will just purge everything
as it's directly reading the source files, seeing anything vanilla-extract related and not idenifying it as css.

One workaround to this is to insert purgecss as a vite plugin instead of via postcss.
This way purgecss can look at the code after it's been transpiled / bundled to extract what css is in use

```
pnpm add -D @mojojoejo/vite-plugin-purgecss
```

`vite.config.ts`
```ts
import { defineConfig } from 'vite'
import pluginPurgeCss from "@mojojoejo/vite-plugin-purgecss"
import vue from '@vitejs/plugin-vue'

export default defineConfig({

  // ...

  plugins: [
    vue(),
    vanillaExtractPlugin({
      identifiers: 'debug'
    }),
    pluginPurgeCss(),
  ],

  // ...

})
```

This does seem to work but has one limitation.
When calling the sprinkles function with a list of styles you have to make sure it's done within a .css.ts file
If it's done directly within the control then purgecss will pick up on all the css props within the final js bundle which is not what you want (unless you're trying to use sprinkles at runtime dynamically)

`Demo1.vue.css.ts`
```ts
import { spr } from './some/sprinkles/definition'
export const style = spr({color: 'rose-300'})
```

`Demo1.vue`
```xml
<template>
  <div :class="style">
    Active
  </div>
  <slot />
</template>

<script setup lang="ts">
// This works
import { style } from './Demo1.vue.css'

// This doesn't and results in all the sprinkles props being brought in
// import { spr } from './some/sprinkles/definition'
// const style = spr({color: 'rose-300'})

</script>
```

**Comment by ilamuhil0907** [comment]
The package you are using `@fullhuman/postcss-purgecss` offers a safelist property along with content 

```
safelist: () => ({
     standard: [/^CLASSPREFIX/, 'myclass']
})
```

Which will allow you to safeguard classes that should not be purged
